### DISCUSSION TITLE
how to write sql to get same result as supabase query method

### DISCUSSION THREAD

**Discussion by atiframlee** [supabase/supabase#2338] [Questions]
in my db i have business,category and product table. through the supabase query method im able to select those table like this:

const { data, error } = await supabase
    .from('business')
    .select(` *,category(*, product(*))`)

With that i manage to return table "business" , containing array of table category, and each category containing table of product. What i seek to do is how to write same query like above via psql which i want to call via psql function.
Thank you

**Comment by ChronSyn** [comment]
Without seeing your schema, it's hard to identify how the tables are linked or what the data structure of each looks like, which makes it very difficult to put together an actual answer.

If you're able to post the schema for those 3 tables (including any foreign key links between tables), some sample data, and the expected output, it'll be a lot easier to put together an answer.

You may be able to do something like this, but again, without more data, it's difficult to provide a comprehensive answer:

```sql
SELECT * FROM
    public.business AS business
LEFT JOIN
    public.product AS product
ON business.product_id
LEFT JOIN
    public.category AS category
ON product.category_id
```

**Reply by atiframlee** [reply]
i created all my table using the supabase interface, im new to sql but i'll try to list my table below as best as i can which contain:

business(
 business_id int4 primary key
 business_name varchar
 business_address varchar
);

each business have category  table of product listed as below -->
category(
 category_id int4 primary key
 business_id int4 references business(business_id)
category_name varchar
 );

and each category have product table listed as below -->
product(
 product_id int4 primary key
 category_id int4 references category(category_id)
 product_name varchar
 product_price float
 product_details
 );

hope this might help to get a better picture at my schema and how to implement the query i seek to.

**Comment by ChronSyn** [comment]
First, a few suggestions/mandatory changes that will absolutely make life easier:

 - Instead of having `category.business_id`, change it to `category.belongs_to_business`. This makes is **significantly** easier to write a query for (as you don't have to deal with ambiguous columns)
 - Likewise, change `product.category_id` to `product.belongs_to_category` for the same reason

Having the same name in multiple tables can be a nightmare to deal with, and you'd have to write significantly more code in your query to work around it. By naming them differently, it's also likely to reduce the chance that a developer will make a mistake - e.g. referencing `category_id` from the wrong table when constructing queries or writing code for your backend or frontend.

I'd also suggest changing `product_price` to be `int4`. You're only removing the decimal, but you're then dealing with the smallest denomination (e.g. instead of 24 dollars and 99 cents, you're dealing with 2499 cents instead). It also prevents situations where certain languages will omit everything after the decimal if it's zero. For example, JS will replace `24.00` with `24`, but `2400` will still be `2400`. Formatting of it can be done further down your stack, perhaps on your front-end.

I did create the query and results based around using a float/real in the event there's a specific reason you want to use them, but I'd still recommend changing to int4.

I had a few troubles getting a fresh Supabase project to behave as expected with int4 columns, so I opted for UUID's instead just to get the query built, but there's no reason the query below wouldn't work for you if you change the relevant `uuid` references to be `int4`:

```sql
CREATE OR REPLACE FUNCTION public.get_products_for_business(business_id_to_find uuid)
RETURNS
	TABLE(
		product_id uuid,
		product_name varchar,
		product_price real,
		category_name varchar,
		category_id uuid,
		business_name varchar,
		business_address varchar,
		business_id uuid
	)
AS $BODY$
BEGIN
  RETURN QUERY
  	SELECT
  		a.product_id,
		a.product_name,
		a.product_price,
		a.category_name,
		a.category_id,
		a.business_name,
		a.business_address,
		a.business_id
	FROM (
		SELECT *
		FROM
			public.product AS products
		LEFT JOIN
			public.category AS categories
			ON products.belongs_to_category = categories.category_id
		LEFT JOIN
			public.business AS businesses
			on categories.belongs_to_business = businesses.business_id
		WHERE businesses.business_id = business_id_to_find
	) a;
END;
$BODY$ LANGUAGE 'plpgsql';


-- -- USAGE EXAMPLE
-- Pass a business_id as the argument
SELECT * FROM get_products_for_business('b9fb9c7d-5973-47cd-a0ce-95b066663bbf')
```

Example output:
![image](https://user-images.githubusercontent.com/12896049/125357781-ffec3300-e35f-11eb-8beb-179458a9c157.png)

The reason why it has a nested query is that without it, it may still complain about ambiguous columns, so we pull out specific properties in the outer query. This also removes columns which have duplicate data, as you don't need the `belongs_to_business` column in the results since you have `business_id` column there already.

**Reply by atiframlee** [reply]
Thank you chronsyn, ur answer really gives me a head start ho top do so.
Bless you


to get the very same result as supabase select query method i had to do it like this

select
business_name,
business_description,
business_admin,
business_address,
business_longitude,
business_latitude,
categories

from business

join (
  select 
  business_fk,
  jsonb_agg( json_build_object('category_id',category_id,'category_name',category_name) ) 
  as categories
  from category

join(
  select category_fk,
  jsonb_agg( jsonb_build_object('product_name',product_name,'product_description',product_description,'product_price',product_price) ) as products
  from product
  group by category_fk
) newproduct on category.category_id = newproduct.category_fk

 group by business_fk
) newcategory on business.business_id = newcategory.business_fk

and now im able to implement it as i seek to. Thanks again
